CI: add Github workflow for KinD testing and static checks

mayankshah1607 opened the pull request:
Fixes https://github.com/linkerd/linkerd2-conformance/issues/4

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add scripts and workflows

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Trigger CI

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update workflow to trigger on push

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update static checks to run on push

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Try help/kind-action

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Switch back to engineerd/setup-kind

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
try upgrading from 2.7.1

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Skip ingress tests on KinD

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati commented on .github/config.yaml:
```diff
@@ -0,0 +1,25 @@
+linkerdVersion: stable-2.8.0
```
Just wondering why we can't use the default way of running tests i.e without a config file? :thinking:

mayankshah1607 replied:
For the only reason that currently we want to skip the ingress tests in the CI - because KinD does not yet support Load balancers and having an external IP. Till we find a better alternative, we need to skip ingress by explicitly mentioning it in a config.
But I think we can remove the remaining settings and simply use the defaults

Pothulapati commented on config.yaml:
```diff
@@ -25,7 +25,7 @@ testCase:
         reinstall: false 
         uninstall: true
-        upgradeFromVersion: stable-2.7.0
+        upgradeFromVersion: stable-2.7.1
```
Good Find!

Pothulapati commented on .github/workflows/static_checks.yaml:
```diff
@@ -0,0 +1,23 @@
+    pull_request: {}
+    push: {}
+jobs:
+    go_lint:
```
Wondering if we could use https://github.com/golangci/golangci-lint directly, so that we don't have to take the hassle of maintaining all these tools induvidually :thinking:

mayankshah1607 replied:
Ah, that makes sense! I think we should directly go ahead with the action it provides - https://golangci-lint.run/usage/install/#ci-installation

mayankshah1607 committed:
Use golangci/golangci-lint-action

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati commented on .github/workflows/kind.yaml:
```diff
@@ -0,0 +1,16 @@
+          mv .github/config.yaml .
+    - name: Running conformance tests in KinD cluster
+      run: |
+          go test -timeout 1h -ginkgo.v
```
Wondering how this would work, for the ingress test-case as there is no externalIP being created? :thinking:

mayankshah1607 replied:
As discussed over Slack, we have a custom `config.yaml` for KinD, so that we can skip ingress on kind - https://github.com/linkerd/linkerd2-conformance/blob/95c1e3b0c192e5f655cc6f50436841a7c4776004/.github/config.yaml#L22

Pothulapati commented on .github/workflows/static_checks.yaml:
```diff
@@ -0,0 +1,24 @@
+              uses: golangci/golangci-lint-action@v1
+              with:
+                  version: v1.29
+    go_format:
```
Can you confirm if `fmt` is not part of the `golangci-lint-action` so that we can remove this if its already there. Also can you remove the `./bin/goimports` as we are not using that anymore?

mayankshah1607 replied:
Yep! `fmt` can be enabled in the CI using a flag. I've made the required changes and cleaned up the `bin/` folder. 😄

mayankshah1607 committed:
Update static checks

- Remove linter / fmt binaries
- Enable fmt in actions

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 commented:
Also, it looks like the linter is failing due to linting errors in the code. I should probably make a refactor that fixes all of this along with test flakiness and assertions once this is merged!

Pothulapati commented:
@mayankshah1607 Yep, just saw this. Can you update the code here itself to fix the linters, so that we don't have checks failing after this is merged? Also, helps us keeping the refactor PR a separate one!

Pothulapati commented:
One last Nit, I see a weird `./target/bin/goimports-` which can be deleted? :thinking:

mayankshah1607 committed:
Remove target

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 commented:
>One last Nit, I see a weird ./target/bin/goimports- which can be deleted

Yep, did not notice that. Deleted, and added to `.gitignore` :)

>  Can you update the code here itself to fix the linters

Sure! :)

mayankshah1607 committed:
Fix linting errors

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati reviewed (approved):
LGTM!

ihcsim reviewed (approved):
LGTM. Once we merge the ingress PR, we can remove the duplicated `.github/config.yaml`, right?
